### ISSUE TITLE
einsum elementwise addition feature request

### ISSUE THREAD

**Issue by mattharrigan** [numpy/numpy#8139]
This is a follow up to my stack overflow [question](http://stackoverflow.com/questions/39976383/gemm-using-numpy-einsum).  I propose adding elementwise addition and subtraction functionality to einsum.  The API would remain pretty mimimal, with the just addition of the + and - characters.  My gemm question would become `D = np.einsum(',ij,jk+,ij', alpha, A, B, beta, C)`.

The parsing/API would be that on each side of a + or -, the operations must result in an array of equal shape which is the added or subtracted elementwise.  A  very simple un-optimized implementation could parse the above as `D = np.einsum(',ij,jk', alpha, A, B) + np.einsum(',ij', beta, C)`

Many optimization opportunities exist, including most of the one mentioned in this [NEP](http://docs.scipy.org/doc/numpy/neps/deferred-ufunc-evaluation.html).  The example problem in the beginning could be `A = np.einsum('...+...+...', B, C, D)`.  It could be optimized to a single pass without intermediate arrays.  It might be easier to implement those optimizations inside einsum, but I don't understand the internals nearly well enough to comment intelligently.  Maybe `D = np.einsum('i,ij->ij+ij', A[0,:], B, C)` could use fused multiply add

Even without clever optimizations, I think the additional functionality would have significant benefits, making the most powerful linear algebra tool I know of even better and more complete.

I am not sure if there are any issues with the API I naively suggested above, but thought it was worth suggesting.  Thank you

**Comment by shoyer** [comment]
It's not obvious to me what a signature like `',ij,jk+,ij'` means -- the postfix use of `+` is rather unusual. Replacing commas with operators and using standard infix notation like `...*ij*jk+ij` would feel much more obvious to me. Then you could even add the full suite of numeric operations and parentheses for explicit precedence.

That said, I don't think this is a good fit for `einsum`, because the syntax looks quite different (recall that einsum also supports an interweaved notation for supplying arguments). I would rather start with a new function name. This could certainly be prototyped/developed in pure Python if you are interested. You can find the parsing logic for `einsum` in `numpy/core/einsumfunc.py`.

**Comment by shoyer** [comment]
Also, before you get too far along this path, keep in mind that all API changes to NumPy need to gain consensus on the [mailing list](https://mail.scipy.org/mailman/listinfo/numpy-discussion).

**Comment by mattharrigan** [comment]
I don't understand why its postfix notation.  The "+" is between its operands.

**Comment by juliantaylor** [comment]
hm isn't einsum mostly used for level 2 and 3 blas operations? is the performance of level 1 operations in this context really relevant in practice?

**Comment by shoyer** [comment]
> I don't understand why its postfix notation. The "+" is between its operands.

You're right, postfix is the wrong word. It does look weird to me with the comma, though.

To be clear, I do like the idea of specifying arbitrary broadcasting arithmetic using similar syntax (without the commas).

**Comment by mattharrigan** [comment]
I don't know how large any performance gains are or how useful they are to the community at large.  I think einsum should support elementwise addition mostly for completeness.  It can already support matrix and elementwise multiplication and reduction by sum.  Supporting elementwise addition would  complete most of the common core linear algebra operations in blas 1 through 3 with a (hopefully still) very elegant syntax.  Performance gains are also nice though.

In case its helpful, here is another example of what I am proposing:
A = np.einsum('ij,jk+ij,jk', B, C, D, E)
where the inputs are 2D arrays of compatible dimensions.  I think the meaning is clear if you already know a bit about the current einsum API.

**Comment by eric-wieser** [comment]
Here's one possible way of supporting index notation

    i, j, k = np.einindices(3)
    A = np.einsum(lambda i, j, k: B[i,j] * C[j,k] + D[i,j]*E[j,k])

In the case where the result is not a scalar, perhaps:

    A = np.einsum(lambda i, j, k, res: res[i,j], B[i,j] * C[j,k] + D[i,j]*E[j,k])

This could be implemented either by:

 * passing in objects for which `ndarray.__getitem__` has a special handler
 * using `ast` to parse a expression tree

**Comment by MichaelSnowden** [comment]
Any suggested workaround for the lack of this feature? I currently have a problem where I need to do something along the lines of either `np.einsum(lambda ...)` or `np.einsum(..., add_to_out=x)`. It's a [pretty complicated question](https://stackoverflow.com/questions/45580013/pure-numpy-2d-mean-convolution-derivative-of-input-image), but if I had an `einsum` where I could either specify the operation done or an `add_to_out` argument, then it would be solved.

**Comment by zerothi** [comment]
add_to_out would be awesome!

**Comment by tpchuckles** [comment]
Just piping in with another example:
the math would be: 
`sum_a{ exp(i 2 pi ( kx xs + ky ys )) }`
where kx is length 'k', xs is length 'a', ky is length 'q', ys is length 'a' (kx,ky are a grid in reciprocal space. you have lists of x,y values for a series of points, we're "painting in" sinusoids for each point).
with broadcasting to 'a,k,q' indices then collapsing along 'a':
`result = np.sum( np.exp(1j*2*np.pi*(kx[None,:,None]*xs[:,None,None]+ky[None,None,:]*ys[:,None,None])) , axis = 0)`
or noting:
`exp(i 2 pi ( kx xs + ky ys )) = exp(i 2 pi kx xs) * exp(i 2 pi ky ys )`
we can minimize broadcasting to a degree with, using einsum with 'a,k' and 'a,q' --> 'k,q':
`expx = np.exp(1j*2*np.pi*(kx[None,:]*xs[:,None])`
`expy = np.exp(1j*2*np.pi*(ky[None,:]*ys[:,None])`
`result = np.einsum('ak,aq->kq',expx,expy)`
but this is still pretty memory-intensive when length along 'a' is large. 

~~what i'd really like is something like:~~
~~`result = np.exp(1j*2*np.pi*np.einsum('k*a+q*a->kq',kx,xs,ky,ys)`~~
~~even if I need to expand to 'k,q' first:~~
~~`result = np.exp(1j*2*np.pi*np.einsum('kq*a+kq*a->kq',kx[:,None],xs,ky[None,:],ys)`~~
~~since 'k,q' is potentially way way way way smaller than either 'a,k' or 'a,q', then it would still be a huge win.~~

~~is there anything here i'm missing? some math magic i didn't notice? or some other way to avoid the intermediate _summed_ 'a,k'+'a,q' which will only be collapsed along 'a'?~~

for reference, we're talking sizes of up to 7500,6500,10000 for 'k','q','a' respectively

edit: 
something like:
`result = np.exp(1j*2*np.pi*(np.einsum('k,a->k',kxs,xs)[:,None]+np.einsum('q,a->q',kys,ys)[None,:]))`
may be tempting and appears identical to the strike-through suggestion above. but we do not want to sum inside the exponent. 

instead, i think i _actually_ want something like this:
`result = np.einsum('e^(k*a+q*a)->kq',kx,xs,ky,ys)`
but this is obviously quite a bit more complicated than simply adding in plus/minus operations...
